### PR TITLE
Orgion/future/card implementaion

### PR DESCRIPTION
**Pull request by purushothamanTR**
card datatable jquery implementaion completd as mvc and api

### CODE REVIEW

File: BookingSystem.DAL/Repository/CardRepository.cs [hunk 1]
```diff
@@ -109,6 +109,78 @@ public async Task<List<CardsVM>> GetActiveCardAsync()
            
         }
 
+        public async Task<CardDatatableVM> GetActiveCardFilterAsync(DatatableParamVM JqueryDatatableParams, int sortColumnIndex
+            , string sortDirection)
+        {
+            var cardInfo = (from c in _dataContext.Cards
+                            join cs in _dataContext.Customer on c.CustomerID equals cs.CustomerID
+                            where c.IsDeleted == false
```
Comment (manic1990) [inline BookingSystem.DAL/Repository/CardRepository.cs:117]: Check card table IsDeleted  false

File: BookingSystem.Models/Model/DatatableParamVM.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace BookingSystem.Models.Model
+{
+    public class DatatableParamVM
+    {
+
+        [DefaultValue("")]
```
Comment (manic1990) [inline BookingSystem.Models/Model/DatatableParamVM.cs:13]: Why we need to set DefaultValue("") i think its always empty string only please check once.

File: BookingSystem/APIClient/Repository/APIClientCardRepository.cs [hunk 3]
```diff
@@ -74,6 +74,32 @@ public async Task<List<CardsVM>> GetActiveCardAsync()
             return default(List<CardsVM>);
         }
 
+        public async Task<CardDatatableVM> sorting(ViewDatatableParam JqueryDatatableParams, int sortColumnIndex, string sortDirection)
+        {
+            var uri = new Uri(httpClient.BaseAddress + Constant.CardDatablestring + "?sortColumnIndex=" + sortColumnIndex + "&sortDirection=" + sortDirection);
```
Comment (manic1990) [inline BookingSystem/APIClient/Repository/APIClientCardRepository.cs:79]: add same constant don't hard code it "?sortColumnIndex=".

File: BookingSystem/Controllers/CardController.cs [hunk 4]
```diff
@@ -53,20 +58,10 @@ public async Task<IActionResult> Card(int Card)
                 {
                     carddetails = _mapper.Map<ViewCardsData>(await _cardRepository.GetActiveCardByIdAsync(Card).ConfigureAwait(false));
                 }
```
Comment (manic1990) [inline BookingSystem/Controllers/CardController.cs:60]: Use try catch block and logger information refer other controller

File: BookingSystem/Controllers/CardController.cs [hunk 5]
```diff
@@ -150,5 +145,85 @@ public async Task<IActionResult> Card(ViewCardsData model)
         }
 
 
+
+        public async Task<IActionResult> GetData(ViewDatatableParam JqueryDatatableParam)
+        {
+          
+
+            var sortColumnIndex = Convert.ToInt32(HttpContext.Request.Query["iSortCol_0"]);
+            var sortDirection = HttpContext.Request.Query["sSortDir_0"];
```
Comment (manic1990) [inline BookingSystem/Controllers/CardController.cs:154]: Line no 153 and 154 add it inside try block

File: BookingSystem/Views/Card/Card.cshtml [hunk 6]
```diff
@@ -92,24 +92,68 @@
...
+                //var disco3 = new RegExp('^6[45]$');
 
+                //var diners = new RegExp('^3[0689]$');
+                //var jcb = new RegExp('^35[0-9]$');
 
+                //if (mastercard.test(VAL)) {
+                //    $("#CardType option[value=1]").attr('selected', 'selected');
+
+
+                //}
```
Comment (manic1990) [inline BookingSystem/Views/Card/Card.cshtml:148]: remove unwanted codes

File: BookingSystem/Views/Card/CardDetails.cshtml [hunk 7]
```diff
@@ -4,59 +4,161 @@
...
+                                            //                    "autoWidth": true,
+                                            //                    "searchable": true
+                                            //                }, {
+                                            //                    "data": "ExpiryYear",
+                                            //                    "autoWidth": true,
+                                            //                    "searchable": true
+                                            //                },
+                                            //            ]
+                                            //        });
+                                            //}
```
Comment (manic1990) [inline BookingSystem/Views/Card/CardDetails.cshtml:159]: remove commented lines

File: BookingSystem.DAL/Repository/CardRepository.cs (hunk 1 above)
Comment (purushothamanTR) [reply BookingSystem.DAL/Repository/CardRepository.cs:117]: isdeleted handled

File: BookingSystem.Models/Model/DatatableParamVM.cs (hunk 2 above)
Comment (purushothamanTR) [reply BookingSystem.Models/Model/DatatableParamVM.cs:13]: swager is act as empty "string" so handle ""

File: BookingSystem/APIClient/Repository/APIClientCardRepository.cs (hunk 3 above)
Comment (purushothamanTR) [reply BookingSystem/APIClient/Repository/APIClientCardRepository.cs:79]: constant added

File: BookingSystem/Controllers/CardController.cs (hunk 4 above)
Comment (purushothamanTR) [reply BookingSystem/Controllers/CardController.cs:60]: insdide try catch block added

File: BookingSystem/Controllers/CardController.cs (hunk 5 above)
Comment (purushothamanTR) [reply BookingSystem/Controllers/CardController.cs:154]: insdide try catch block added

File: BookingSystem/Views/Card/Card.cshtml (hunk 6 above)
Comment (purushothamanTR) [reply BookingSystem/Views/Card/Card.cshtml:148]: removed

File: BookingSystem/Views/Card/CardDetails.cshtml (hunk 7 above)
Comment (purushothamanTR) [reply BookingSystem/Views/Card/CardDetails.cshtml:159]: remove commented lines

File: BookingSystem/APIClient/Interface/IAPIClientCardRepository.cs [hunk 8]
```diff
@@ -5,10 +5,12 @@
...
 namespace BookingSystem.APIClient.Interface
 {
-    public interface  IAPIClientCardRepository
+    public interface IAPIClientCardRepository
     {
         Task<CardsVM> SaveCardAsync(CardsVM cards);
         Task<CardsVM> GetActiveCardByIdAsync(int CardID);
         Task<List<CardsVM>> GetActiveCardAsync();
+
+        Task<CardDatatableVM> sorting(ViewDatatableParam JqueryDatatableParams,int sortColumnIndex,string sortDirection);
```
Comment (sathish199308) [inline BookingSystem/APIClient/Interface/IAPIClientCardRepository.cs]: Kindly change the method name as Suffix Async- SortingAsync, Check if it input parameters is coming null whether it throwing error or not

File: BookingSystem/Constants/Constant.cs [hunk 9]
```diff
@@ -33,5 +33,14 @@ public static class Constant
         public const string Get_Customer_By_Id = "customer/getcustomerbyId";
         public const string Delete_Customer_By_Id = "customer/deletecustomerbyId";
         public const string Get_Booking_Async = "booking/getbookingasync";
+        public const string CardDatablestring = "card/GetActiveCardFilterAsync";
+        public const string sortColumnIndex = "sortColumnIndex";
```
Comment (sathish199308) [inline BookingSystem/Constants/Constant.cs]: Check the variable name either we follow Pascalcase or CamelCase

File: BookingSystem/Views/Card/Card.cshtml [hunk 10]
```diff
@@ -92,24 +92,68 @@
...
+
+            //$('.cardnumber').keyup(function () {
+            let VAL = this.value;
+
+            if (VAL.length == 4) {
+
+                if (VAL.match(/^5[1-5]/)) {
+                    $("#CardType option[value=1]").attr('selected', 'selected');
+                }
+                //} else if (num.match(/^4\d{15}/) || num.match(/^4\d{12}/)) {
```
Comment (sathish199308) [inline BookingSystem/Views/Card/Card.cshtml:118]: Remove the comments line

**Review by sathish199308 — APPROVED** [review]
Approved

**Review by sathish199308 — APPROVED** [review]
Constant File Updated change is Approved

**Comment by sathish199308** [comment]
DataTable Integration changes is merged

**Pull request merged**